README.md
# Flutter Spelling Number
 Flutter Spelling Number it's a plugin convert number to words - their written form(تفقيط أوتحويل الأرقام الى كلمات).

## Examples
### [Convert Number to words](https://mattar88.github.io/flutter_spelling_number/example#spelling_en_number)
### [Convert Dollars and Cents into words](https://mattar88.github.io/flutter_spelling_number/example#spelling_en_money)
### [Convert Weight number into words](https://mattar88.github.io/flutter_spelling_number/example#spelling_en_weight)
### [تفقيط أو تحويل الرقم الى كلمات](https://mattar88.github.io/flutter_spelling_number/example#spelling_ar_number)
### [تفقيط مبالغ مالية وعملات](https://mattar88.github.io/flutter_spelling_number/example#spelling_ar_money)
### [تفقيط الوزن](https://mattar88.github.io/flutter_spelling_number/example#spelling_ar_weight)

## Features:
 *  Support 14 languages includes arabic language
 *  You can specify the Whole unit and Fraction Unit and the Decimal separator
 *  you can convert Currency and Weight number to words
 
## Install

With Dart:
```
$dart pub add spelling_number
```
With Flutter:
```
$flutter pub add spelling_number
```
This will add a line like this to you package's pubspec.yaml (and run an implicit `dart pub get`):
```
dependencies:
  spelling_number:^0.0.4
```

## Usage
 
Import it
Now in your Dart code, you can use:

 ```
 import 'package:spelling_number/spelling_number.dart';
 ```

 ```
     SpellingNumber( 
         lang: "en",
        wholesUnit: "ton",
        fractionUnit: "kg",
        digitsLengthW2F: 3,
        decimalSeperator: "and").convert(100.951); //=> One Hundred Ton And Nine Hundred Fifty-One Kg
  
    SpellingNumber().convert(1234); // => 'one thousand two hundred and thirty-four'
    SpellingNumber(lang: 'es').convert(4758); // => 'cuatro mil setecientos cincuenta y ocho'
    SpellingNumber(lang: 'fr').convert(1234);   // => 'mille deux cent trente-quatre'
    SpellingNumber(lang: 'es').convert(1234);   // => 'mil doscientos treinta y cuatro'
    SpellingNumber(lang: 'az').convert(1234);   // => 'min iki yüz otuz dörd'
    SpellingNumber(lang: 'pt').convert(1234);   // => 'mil duzentos e trinta e quatro'
    SpellingNumber(lang: 'ar').convert(1234);   // => 'ألف ومائتان وأربعة وثلاثون'
    SpellingNumber(lang: 'eo').convert(1234);   // => 'mil ducent tridek kvar'
    SpellingNumber(lang: 'vi').convert(1234);   // => 'một ngàn hai trăm và ba mươi bốn'
    SpellingNumber(lang: 'uk').convert(1234);   // => 'одна тисяча двісті тридцять чотири'
    SpellingNumber(lang: 'id').convert(1234);   // => 'seribu dua ratus tiga puluh empat'
   ```

## Constructor parameters
- `noAnd` - Defaults to `false`. Determines whether to use a separator. The
  separator is internationalized.
- `lang` - Could be `string` or `object`. Defaults to `'en'`. Determines which
  language to use. An i18n configuration object may be passed to support
  external language definitions.

Currently supported languages are:

| Language | `lang` |
|---------|--------|
| English | `en` |
| Portuguese (Brazil) | `pt` |
| Portuguese (Portugal) | `ptPT` |
| Spanish | `es` |
| French | `fr` |
| Esperanto | `eo` |
| Vietnamese | `vi` |
| Arabic | `ar` |
| Azerbaijan | `az` |
| Turkish | `tr` |
| English (Indian) | `enIndian` |
| Ukrainian | `uk` |
| Indonesian | `id` |
| Russian | `ru` |


## Contributing

### Configure your own language
Each language has its own unique grammar exceptions.  You can create your own 
language.json file in the folder "i18n" and give writtenNumber support for it. I 
don't think the current scheme and logic cover all the cases, but may be cover 
some.

The following parameters have been used for the currently available languages:


### Language parameters

| Parameter       | Type    | Description                                                                                                                                        | Examples                                                                                                                                                                     |
|-----------------|---------|----------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `useLongScale`    | boolean | Indicates if it uses [long or short scale](https://en.wikipedia.org/wiki/Long_and_short_scales).                                                    | This differs the meaning of the words `billion`, `trillion` and so on.                                                                                                       |
| `baseSeparator`   | string  | Separates the base cardinal numbers.                                                                                                               | 29 -> twenty`-`eight.  Spanish uses the connector " y "                                                                                                                      |
| `unitSeparator`   | string  | Separates the units from the last base cardinal numbers.                                                                                           | 1234 -> one thousand two hundred **and** thirty-four                                                                                                                         |
| `allSeparator`    | string  | Separates all cardinals, not only the last one.                                                                                                    | 1125 -> ألف **و**مائة **و**خمسة **و**عشرون                                                                                                                                   |
| `base`            | Object  | Base cardinals numbers.  Numbers that have unique names and are used to build others.                                                              |                                                                                                                                                                              |
| `alternativeBase` | Object  | Alternative versions of base cardinals numbers for usage with specific units.  These bases will be treated as an extension for the default `base`. | ``` "alternativeBase": {   "feminine": {"1":"одна","2":"дві"} } ```                                                                                                          |
| `units`           | Array   | A list of number units (string or Object). Gives support to singular, dual an plural units. Check the Object parameters below.                                                |                                                                                                                                                                              |
| `unitExceptions`  | Object  | Sometimes grammar exceptions affect the base cardinal joined to the unit. You can set specific exceptions to any base cardinal number.             | Converting 1232000 in Spanish:  Without Exception (Wrong):  -> **uno** millón doscientos treinta y dos mil  With Exception:  -> **un** millón doscientos treinta y dos mil   |

### Units parameters

A `unit` can be:
- A simple string. e.g. `"hundred"`
- An Object with multiple parameters:

| Unit parameter         | Description                                                                                            | e.g. of languages |
|------------------------|--------------------------------------------------------------------------------------------------------|-------------------|
| `singular`             | One element.                                                                                           | All               |
| `dual`                 | Two elements.                                                                                          | `ar`              |
| `plural`               | Two or more elements. (or 3 or more)                                                                   | All               |
| `few`                  | Between 2 and 4 including.                                                                             | `uk`              |
| `useAlternativeBase`   | Overwrites default `base`.                                                                             | `uk`              |
| `useBaseInstead`       | Use the base cardinal number instead.                                                                  | `es`,`hu`,`pt`    |
| `useBaseException`     | Specify with which unit (1 to 9) you don't want to use the base, and instead use the regular behavior. | `es`,`hu`,`pt`    |
| `avoidPrefixException` | Units not using the base cardinal number prefix for unit 1.                                            | `id`,`tr`,`it`    |
| `avoidInNumberPlural`  | Units not using the plural form with trailing numbers other than 0.                                    | `fr`              |
| `restrictedPlural`     | Plural only for 3 to 10. Singular if >= 11.                                                            | `ar`              |
| `useSingularEnding`    | Use singular form for numbers ending with 1.                                                           | `uk`              |
| `useFewEnding`         | Use _few_ form for numbers ending with 2, 3 or 4.                                                      | `uk`              |
| `avoidEndingRules`     | Plural form used instead of `useSingularEnding` and `useFewEnding`                                     | `uk`              |


CHANGELOG.md
## 0.0.1

* First publish.

## 0.0.2

* Improve documentation.

## 0.0.3

* Improve code and documentation.

## 0.0.4

* Improve code and documentation.



test/spelling_number_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:spelling_number/spelling_number.dart';

void main() {
  test(
      'Run 2 tests the first convert number to words the second convert money number to words',
      () {
    expect(
        SpellingNumber(
          lang: 'en',
          digitsLengthW2F: 2,
          decimalSeperator: "and",
        ).convert(132.21),
        "one hundred and thirty-two  and twenty-one");

    expect(
        SpellingNumber(
          lang: 'en',
          wholesUnit: "dollar",
          fractionUnit: "cent",
          digitsLengthW2F: 2,
          decimalSeperator: "and",
        ).convert(1123000530.12),
        "one billion one hundred twenty-three million five hundred and thirty dollar and twelve cent");
  });
}


example/lib/main.dart
import 'dart:developer';

import 'package:flutter/material.dart';
import 'package:spelling_number/spelling_number.dart';
import 'dart:math' as Math;

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Spelling Number'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<Map<String, dynamic>> spelledNumberList = [];

  @override
  void initState() {
    var numberFormats = [
      {
        'lang': 'en',
        'digitsLengthW2F': 2,
        'decimalSeperator': "and",
        'header': 'Example Convert Number into words'
      },
      {
        'lang': 'ar',
        'digitsLengthW2F': 2,
        'decimalSeperator': "و",
        'header': ' تفقيط أو تحويل الرقم الى كلمات باللغة العربية'
      },
    ];

    var numbers = [100, 132.21, 1000000, 1123000530.12];

    for (var numberFormat in numberFormats) {
      spelledNumberList
          .add({'title': numberFormat['header'].toString(), 'value': ""});
      for (int i = 0; i < numbers.length; i++) {
        spelledNumberList.add({
          'title':
              'The spelling of ${numbers[i]} in ${numberFormat['lang']} lang is',
          'value': SpellingNumber(
            lang: numberFormat['lang'].toString(),
            digitsLengthW2F: numberFormat['digitsLengthW2F'] as int,
            decimalSeperator: numberFormat['decimalSeperator'].toString(),
          ).convert(numbers[i])
        });
        // log('The spelling of ${numbers[i]} in language ${numberFormat['lang']} is : ${SpellingNumber(
        //   lang: numberFormat['lang'].toString(),
        //   digitsLengthW2F: numberFormat['digitsLengthW2F'] as int,
        //   decimalSeperator: numberFormat['decimalSeperator'].toString(),
        // ).convert(numbers[i])}');
      }
    }

    var moneyFormats = [
      {
        'lang': 'en',
        'wholesUnit': "dollar",
        'fractionUnit': "cent",
        'digitsLengthW2F': 2,
        'decimalSeperator': "and",
        'header': 'Example Convert Dollars and Cents into words'
      },
      {
        'lang': 'ar',
        'wholesUnit': "ليرة لبنانية",
        'fractionUnit': "قرش",
        'digitsLengthW2F':
            3, //The pound or lira is the currency of Lebanon. It was formerly divided into 100 piastres
        'decimalSeperator': "و",
        'header': 'تفقيط العملة'
      },
    ];

    for (var format in moneyFormats) {
      spelledNumberList
          .add({'title': format['header'].toString(), 'value': ""});
      for (int i = 0; i < numbers.length; i++) {
        spelledNumberList.add({
          'title':
              'The spelling of amount ${numbers[i]} in ${format['lang']} lang is',
          'value': SpellingNumber(
            lang: format['lang'].toString(),
            wholesUnit: format['wholesUnit'].toString(),
            fractionUnit: format['fractionUnit'].toString(),
            digitsLengthW2F: format['digitsLengthW2F'] as int,
            decimalSeperator: format['decimalSeperator'].toString(),
          ).convert(numbers[i])
        });
        // log('The spelling amount of ${numbers[i]} is : ${SpellingNumber(
        //   lang: format['lang'].toString(),
        //   wholesUnit: format['wholesUnit'].toString(),
        //   fractionUnit: format['fractionUnit'].toString(),
        //   digitsLengthW2F: format['digitsLengthW2F'] as int,
        //   decimalSeperator: format['decimalSeperator'].toString(),
        // ).convert(numbers[i])}');
      }
    }

    var weightFormats = [
      {
        'lang': 'en',
        'wholesUnit': "Ton",
        'fractionUnit': "Kg",
        'digitsLengthW2F': 3,
        'decimalSeperator': "and",
        'header': 'Example Convert Weight number into words'
      },
      {
        'lang': 'ar',
        'wholesUnit': "طن",
        'fractionUnit': "كيلو غرام",
        'digitsLengthW2F':
            3, //The pound or lira is the currency of Lebanon. It was formerly divided into 100 piastres
        'decimalSeperator': "و",
        'header': 'تفقيط الوزن'
      },
    ];

    for (var weightFormat in weightFormats) {
      spelledNumberList
          .add({'title': weightFormat['header'].toString(), 'value': ""});
      for (int i = 0; i < numbers.length; i++) {
        spelledNumberList.add({
          'title':
              'The spelling of weight ${numbers[i]} in ${weightFormat['lang']} lang is',
          'value': SpellingNumber(
            lang: weightFormat['lang'].toString(),
            wholesUnit: weightFormat['wholesUnit'].toString(),
            fractionUnit: weightFormat['fractionUnit'].toString(),
            digitsLengthW2F: weightFormat['digitsLengthW2F'] as int,
            decimalSeperator: weightFormat['decimalSeperator'].toString(),
          ).convert(numbers[i])
        });
      }
    }
    log('${spelledNumberList}');

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Container(
        // decoration: const BoxDecoration(color: Color(0xff121b22)),
        child: ListView.builder(
            itemCount: spelledNumberList.length,
            itemBuilder: (context, index) {
              var item = spelledNumberList[index];
              if (item['value'].length == 0) {
                return Center(
                  child: Container(
                    width: double.infinity,
                    height: 30,
                    child: Text(
                      item['title'].toString(),
                      style: TextStyle(
                          fontWeight: FontWeight.bold, color: Colors.white),
                    ),
                    color: Colors.blue,
                    alignment: AlignmentDirectional.center,
                  ),
                );
              }
              return ListTile(
                title: Text(
                  item['title'].toString(),
                  style: const TextStyle(
                      color: Colors.blue, fontWeight: FontWeight.bold),
                ),
                subtitle: Text(
                  item['value'].toString(),
                  style: const TextStyle(color: Color(0xff738088)),
                ),
              );
            }),
      ),
      // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
